datastore/src/main/java/io/spine/server/storage/datastore/DsLookupByIds.java

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * An {@code Entity} lookup in Google Datastore using {@code Entity} identifiers.
+ *
+ * @implNote Lookup is performed by reading all the entities with Datastore Keys matching
+ *         the provided IDs first and then applying other query constraints in-memory.
+ */
+final class DsLookupByIds<I> {
+
+    private final DatastoreWrapper datastore;
+    private final TypeUrl typeUrl;
+
+    DsLookupByIds(DatastoreWrapper datastore, TypeUrl url) {
+        this.datastore = datastore;
+        this.typeUrl = url;
+    }
+
+    /**
+     * Queries at most the specified amount of records with supplied identifiers which match the
+     * provided predicate, and applies a field mask to the query results.
+     *
+     * <p>The results are returned in an order specified by the provided
+     * {@linkplain io.spine.client.OrderBy order clause}.
+     *
+     * @param ids
+     *         entity identifiers which are translated to Datastore keys
+     * @param fieldMask
+     *         a field mask specifying fields to be included in resulting entities
+     * @param predicate
+     *         a predicate which must be matched by entities to be returned as results
+     * @param orderBy
+     *         a specification of order in which the query results must be returned
+     * @return an iterator over the matching entity records
+     */
+    Iterator<EntityRecord> find(Iterable<I> ids,
+                                FieldMask fieldMask,
+                                Predicate<Entity> predicate,
+                                OrderBy orderBy,
+                                long limit) {
+        return read(ids)
+                .filter(Objects::nonNull)
+                .filter(predicate)
+                .sorted(implementing(orderBy))
+                .map(Entities::toRecord)
+                .map(maskRecord(fieldMask))
+                .limit(limit)
+                .iterator();
+    }
+
+    /**
+     * Queries the records with supplied identifiers which match the provided predicate,
+     * and applies a field mask to the query results.
+     *
+     * <p>The results are returned in an order specified by the provided
+     * {@linkplain io.spine.client.OrderBy order clause}.
+     *
+     * @param ids
+     *         entity identifiers which are translated to Datastore keys
+     * @param fieldMask
+     *         a field mask specifying fields to be included in resulting entities
+     * @param predicate
+     *         a predicate which must be matched by entities to be returned as results
+     * @param orderBy
+     *         a specification of order in which the query results must be returned
+     * @return an iterator over the matching entity records
+     */
+    Iterator<EntityRecord> find(Iterable<I> ids,
+                                FieldMask fieldMask,
+                                Predicate<Entity> predicate,
+                                OrderBy orderBy) {
+        return read(ids)
```

Comment:
Same as above.

I also don't understand why this method and the method above do not share any code. Please review this method and all the similar methods below for returning the variable (useful for inspecting in debug) and for the code duplication.